bricks/distributed/vector_clock.h

```diff
@@ -0,0 +1,130 @@
+    std::vector<std::chrono::microseconds>::resize(size, now);
+  }
+};
+
+template <class CLOCK = DiscreteClocks>
+class MergeStrategy {
+ public:
+  std::pair<bool, bool> merge(CLOCK &v1, CLOCK &v2, std::function<bool(CLOCK &v1, CLOCK &v2)> validator) {
+    // Basic merge strategy:
+    // 1. always merge (even in case of conflict)
+    // 2. default conflict function check that v1 <= v2
+    return std::pair<bool, bool>{!validator(v1, v2), true};
+  }
+  std::pair<bool, bool> merge(CLOCK &v1, CLOCK &v2) { return merge(v1, v2, is_conflicting); }
+  static bool is_conflicting(CLOCK &v1, CLOCK &v2) {
+    // basic implementation
+    // returns true if there are no conflicts for merging
+    return !is_lte(v1, v2);
+  }
+
+  static bool is_same(CLOCK &v1, CLOCK &v2) {
+    // Happens on exactly same moment
+    // T=T' if T[i] = T'[i] for any i
+    for (size_t i = 0; i < v1.size(); i++) {
+      if (v1[i] != v2[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  static bool is_lte(CLOCK &v1, CLOCK &v2) {
+    // Happens early or at the same time
+    // T <= T' if T[i] <= T'[i] for any i
+    for (size_t i = 0; i < v1.size(); i++) {
+      if (v1[i] > v2[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  static bool is_early(CLOCK &v1, CLOCK &v2) {
+    // v1 happens before v2
+    // T < T' if T <= T' and T != T'
+    return is_lte(v1, v2) && !is_same(v1, v2);
+  }
+
+  static bool is_parallel(CLOCK &v1, CLOCK &v2) {
+    // v1 is not in sync with v2
+    // T || T' if !(T <= T') and !(T' <= T)
+    return !is_lte(v1, v2) && !is_lte(v2, v1);
+  }
+};
+
+template <class CLOCK = DiscreteClocks>
+class StrictMergeStrategy : public MergeStrategy<CLOCK> {
+ public:
+  static bool is_conflicting(CLOCK &v1, CLOCK &v2) {
+    // Check if v1 is in sync with v2 and v1 is strictly early then v2
+    return !(!MergeStrategy<CLOCK>::is_parallel(v1, v2) && MergeStrategy<CLOCK>::is_early(v1, v2));
+  }
+  std::pair<bool, bool> merge(CLOCK &v1, CLOCK &v2) { return MergeStrategy<CLOCK>::merge(v1, v2, is_conflicting); }
+};
+
+template <class CLOCK = DiscreteClocks, template <typename> class STRATEGY = MergeStrategy>
+class VectorClock {
+ protected:
+  CLOCK clock;
+  STRATEGY<CLOCK> strategy;
+  uint32_t local_id;
+
+ public:
+  explicit VectorClock(uint32_t size, uint32_t node_id) {
+    // Set local process id and cluster size
+    local_id = node_id;
+    clock.reset(size);
+  }
+
+  explicit VectorClock(CLOCK &v, uint32_t node_id) {
+    // Constructor for existing clock, used for inserting new data
+    local_id = node_id;
+    clock = CLOCK(v.begin(), v.end());
+  }
+
+  explicit VectorClock() {
+    // Lamport clocks for size=1
+    local_id = 0;
+    clock.reset(1);
+  }
+
+  void step() {
+    // T[i] = T[i] + 1 for logical step
+    clock.inc(local_id);
+  }
+
+  CLOCK &state() {
+    // Returns current state for network transmission
+    return clock;
+  }
+
+  bool merge(CLOCK &to_compare) {
```

Comment:
Not really clear from this code what's mutated and when -- I'd say either always mutate or always stay immutable, to have your users write `auto merged = merge(a, b);`, otherwise it's a potential source of bugs.